src/backend/distributed/metadata/node_metadata.c

```diff
@@ -523,27 +526,96 @@ citus_disable_node(PG_FUNCTION_ARGS)
+		 *      - Pros for the sync option:
+		 *          (a) the changes become visible on the cluster immediately
+		 *          (b) even if the first worker node is disabled, there is no
+		 *              risk of divergence of the placements of replicated shards
+		 *      - Cons for the sync options:
+		 *          (a) Does not work within 2PC transaction (e.g., BEGIN;
+		 *              citus_disable_node(); PREPARE TRANSACTION ...);
+		 *          (b) If there are multiple node failures (e.g., one another node
+		 *              than the current node being disabled), the sync option would
+		 *              fail because it'd try to sync the metadata changes to a node
+		 *              that is not up and running.
+		 */
+		if (firstWorkerNode && firstWorkerNode->nodeId == workerNode->nodeId)
+		{
+			/*
+			 * We cannot let any modification query on a replicated table to run
+			 * concurrently with citus_disable_node() on the first worker node. If
+			 * we let that, some worker nodes might calculate FirstWorkerNode()
+			 * different than others. See LockShardListResourcesOnFirstWorker()
+			 * for the details.
+			 */
+			BlockDistributedQueriesOnMetadataNodes();
+		}
+
+		SyncNodeMetadataToNodes();
+	}
+	else if (UnsetMetadataSyncedForAllWorkers())
+	{
+		/*
+		 * We have not propagated the node metadata changes yet, make sure that all the
+		 * active nodes get the metadata updates. We defer this operation to the
+		 * background worker to make it possible disabling nodes when multiple nodes
+		 * are down.
+		 *
+		 * Note that the active placements reside on the active nodes. Hence, when
+		 * Citus finds active placements, it filters out the placements that are on
+		 * the disabled nodes. That's why, we don't have to change/sync placement
+		 * metadata at this point. Instead, we defer that to citus_activate_node()
+		 * where we expect all nodes up and running.
+		 */
+
 		TriggerNodeMetadataSyncOnCommit();
 	}
 
 	PG_RETURN_VOID();
 }
 
 
+/*
+ * BlockDistributedQueriesOnMetadataNodes blocks all the
+ */
+static void
+BlockDistributedQueriesOnMetadataNodes(void)
+{
+	/* first, block on the coordinator */
+	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
+
+	List *workerList = ActivePrimaryNonCoordinatorNodeList(NoLock);
+	WorkerNode *workerNodeToLock = NULL;
+	foreach_ptr(workerNodeToLock, workerList)
+	{
+		if (!workerNodeToLock->hasMetadata)
+		{
+			/* non-metadata workers cannot run distributed queries, so skip */
+			continue;
+		}
+
+		/*
+		 * Note that we might re-design this lock to be more granular than
+		 * pg_dist_node, scoping only for modifications on the replicated
+		 * tables. However, we currently do not have any such mechanism and
+		 * given that citus_disable_node() runs instantly, it seems acceptable
+		 * to block reads (or modifications on non-replicated tables) for
+		 * a while.
+		 */
+		List *commandList =
+			list_make1("LOCK TABLE pg_dist_node IN ACCESS EXCLUSIVE MODE;");
```

Comment:
Should do EXCLUSIVE instead of ACCESS EXCLUSIVE ? (like above, to avoid blocking queries on pg_dist_node / control plane ops)

Reply:
makes sense, changed to EXCLUSIVE / RowExclusive pair.  This also allows reads to continue, which is actually what we want anyway.

Reply:
Note that this change triggered few isolation test changes, which seems CORRECT to me. But would be great to have a second eye to look in detail @marcocitus:

Two class of changes happened:
1) Because we are running on MX, we acquire `LockShardListResourcesOnFirstWorker`, which calls GetFirstPrimaryWorkerNode(). That's why several `citus_add_node` tests with concurrent reference table modifications become blocked, even if ` citus.replicate_reference_tables_on_activate To off`.  This seems like a bugfix. 

2) Create restore point with concurrent reference table modifications now blocks each other. To be honest, I cannot see how it was NOT the case. This also seems like a bugfix to me.


See the 4th commit: https://github.com/citusdata/citus/pull/5912/commits/c53acd1ef3b5fe02d477c79769c4507383ec18a3